File: demo.vic/src/main/kotlin/com/example/kotlin/fitness/demo/vic/service/ProductService.java
package com.example.kotlin.fitness.demo.vic.service;

import com.example.kotlin.fitness.demo.vic.builders.ProductBuilder;
import com.example.kotlin.fitness.demo.vic.entity.Product;
import com.example.kotlin.fitness.demo.vic.entity.User;
import com.example.kotlin.fitness.demo.vic.entity.api.IProductStorage;
import com.example.kotlin.fitness.demo.vic.exception.LockException;
import com.example.kotlin.fitness.demo.vic.idto.InputProductDTO;
import com.example.kotlin.fitness.demo.vic.mappers.ProductMapper;
import com.example.kotlin.fitness.demo.vic.odto.OutPage;
import com.example.kotlin.fitness.demo.vic.odto.OutputProductDTO;
import com.example.kotlin.fitness.demo.vic.security.filter.JwtUtil;
import com.example.kotlin.fitness.demo.vic.service.api.IAuditService;
import com.example.kotlin.fitness.demo.vic.service.api.IProductService;
import com.example.kotlin.fitness.demo.vic.service.api.IUserService;
import com.example.kotlin.fitness.demo.vic.util.EntityType;
import jakarta.persistence.OptimisticLockException;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Clock;
import java.util.UUID;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class ProductService implements IProductService {

    private final String CREATED = "Line in food journal was created";
    private final String UPDATED = "Line in food journal was updated";

    private final String EDITED = "Line already edited by somebody else";

    private final String LOCK = "Editing forbidden";

    @Autowired
    private final IProductStorage storage;
    @Autowired
    private final IUserService userService;
    @Autowired
    private final JwtUtil jwtUtil;
    @Autowired
    private final IAuditService auditService;
    @Autowired
    private final ProductMapper productMapper;


    @Override
    @Transactional
    public Product create(InputProductDTO idto, String header) {

        validate(idto);
        String login = userService.extractCurrentToken(header);
        String mail = SecurityContextHolder.getContext().getAuthentication().getName();
        User user = userService.loadCurrentUserByLogin(mail);


        Product product = storage.save(ProductBuilder
                .create()
                .setId(UUID.randomUUID())
                .setDtCreate(Clock.systemUTC().millis())
                .setDtUpdate(Clock.systemUTC().millis())
                .setTitle(idto.getTitle())
                .setWeight(idto.getWeight())
                .setCalories(idto.getCalories())
                .setProteins(idto.getProteins())
                .setFats(idto.getFats())
                .setCarbohydrates(idto.getCarbohydrates())
                .setCreatedByRole(login)
                .build());

        auditService.create(
                user,
                EntityType.PRODUCT,
                CREATED,
                product.getId().toString()
        );

        return product;
    }

    public Product read(UUID uuid) {
        return storage.findById(uuid).orElseThrow();
    }

    @Override
    public OutPage<OutputProductDTO> get(Pageable pag) {
        Page<Product> pageOfProduct = storage.findAll(pag);
        return productMapper.map(pageOfProduct);
    }

    @Override
    @Transactional
    public Product update(UUID id, Long dtUpdate, InputProductDTO idto, String header) throws LockException {

        validate(idto);
        Product readed = storage.findById(id).orElseThrow();

        String mail = SecurityContextHolder.getContext().getAuthentication().getName();
        User user = userService.loadCurrentUserByLogin(mail);

        if (!readed.getCreatedByRole().equals(user.getLogin())) {
            if (!user.getRole().equals("ROLE_ADMIN")) {
                throw new LockException(LOCK);
            }
        }

        String login = userService.extractCurrentToken(header);

       /* if (readed == null) {
            throw new IllegalArgumentException("Product wasn't found");
        }*/


        if (!readed.getDtUpdate().equals(dtUpdate)) {
            throw new OptimisticLockException(EDITED);
        }

        Product productUpdate = storage.save(ProductBuilder
                .create()
                .setId(readed.getId())
                .setDtCreate(readed.getDtCreate())
                .setDtUpdate(Clock.systemUTC().millis())
                .setTitle(idto.getTitle())
                .setWeight(idto.getWeight())
                .setCalories(idto.getCalories())
                .setProteins(idto.getProteins())
                .setFats(idto.getFats())
                .setCarbohydrates(idto.getCarbohydrates())
                .setCreatedByRole(login)
                .build());


        auditService.create(user, EntityType.PRODUCT, UPDATED, productUpdate.getId().toString());


        return productUpdate;
    }

    @Override
    public void delete(Product product) {
    }

    public void validate(InputProductDTO idto) {
        if (idto == null) {
            throw new IllegalStateException("You didn't pass the product");
        }
        if (idto.getTitle() == null) {
            throw new IllegalStateException("You didn't pass the Title");
        }
        if (idto.getFats() < 0 || idto.getFats() == 0) {
            throw new IllegalStateException("You didn't pass the value of Fats");
        }
        if (idto.getCarbohydrates() < 0 || idto.getCarbohydrates() == 0) {
            throw new IllegalStateException("You didn't pass the value of Carbohydrates");
        }
        if (idto.getProteins()<0 || idto.getProteins() == 0){
            throw new IllegalStateException("You didn't pass the value of Proteins");
        }
        if (idto.getCalories()<0 || idto.getCalories() ==0){
            throw new IllegalStateException("You didn't pass the value of Calories");
        }
        if (idto.getWeight()<0 || idto.getWeight()==0){
            throw new IllegalStateException("You didn't pass the value of Weight");
        }
    }
}

File: demo.vic/src/main/kotlin/com/example/kotlin/fitness/demo/vic/service/ProductService.kt
package com.example.kotlin.fitness.demo.vic.service

import com.example.kotlin.fitness.demo.vic.builders.ProductBuilder.Companion.create
import com.example.kotlin.fitness.demo.vic.entity.Product
import com.example.kotlin.fitness.demo.vic.exception.LockException
import com.example.kotlin.fitness.demo.vic.idto.InputProductDTO
import com.example.kotlin.fitness.demo.vic.mappers.ProductMapper
import com.example.kotlin.fitness.demo.vic.odto.OutPage
import com.example.kotlin.fitness.demo.vic.odto.OutputProductDTO
import com.example.kotlin.fitness.demo.vic.security.filter.JwtUtil
import com.example.kotlin.fitness.demo.vic.service.api.IAuditService
import com.example.kotlin.fitness.demo.vic.service.api.IProductService
import com.example.kotlin.fitness.demo.vic.service.api.IUserService
import com.example.kotlin.fitness.demo.vic.storage.IProductStorage
import com.example.kotlin.fitness.demo.vic.util.EntityType
import jakarta.persistence.OptimisticLockException
import lombok.RequiredArgsConstructor
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Pageable
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Clock
import java.util.*

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
class ProductService : IProductService {
    private val CREATED = "Line in food journal was created"
    private val UPDATED = "Line in food journal was updated"
    private val EDITED = "Line already edited by somebody else"
    private val LOCK = "Editing forbidden"

    @Autowired
    private val storage: IProductStorage? = null

    @Autowired
    private val userService: IUserService? = null

    @Autowired
    private val jwtUtil: JwtUtil? = null

    @Autowired
    private val auditService: IAuditService? = null

    @Autowired
    private val productMapper: ProductMapper? = null
    @Transactional
    override fun create(idto: InputProductDTO, header: String?): Product? {
        validate(idto)
        val login = userService!!.extractCurrentToken(header)
        val mail = SecurityContextHolder.getContext().authentication.name
        val user = userService.loadCurrentUserByLogin(mail)
        val product = storage!!.save(
            create()
                .setId(UUID.randomUUID())
                .setDtCreate(Clock.systemUTC().millis())
                .setDtUpdate(Clock.systemUTC().millis())
                .setTitle(idto.title)
                .setWeight(idto.weight)
                .setCalories(idto.calories)
                .setProteins(idto.proteins)
                .setFats(idto.fats)
                .setCarbohydrates(idto.carbohydrates)
                .setCreatedByRole(login)
                .build()
        )
        auditService!!.create(
            user,
            EntityType.PRODUCT,
            CREATED,
            product.id.toString()
        )
        return product
    }

    override fun read(uuid: UUID?): Product? {
        return storage!!.findById(uuid!!).orElseThrow()
    }

    override fun get(pag: Pageable?): OutPage<OutputProductDTO?>? {
        val pageOfProduct = storage!!.findAll(pag!!)
        return productMapper!!.map(pageOfProduct)
    }

    @Transactional
    @Throws(LockException::class)
    override fun update(id: UUID?, dtUpdate: Long?, idto: InputProductDTO, header: String?): Product? {
        validate(idto)
        val readed = storage!!.findById(id!!).orElseThrow()!!
        val mail = SecurityContextHolder.getContext().authentication.name
        val user = userService!!.loadCurrentUserByLogin(mail)
        if (readed.createdByRole != user!!.login) {
            if (user.role != "ROLE_ADMIN") {
                throw LockException(LOCK)
            }
        }
        val login = userService.extractCurrentToken(header)

        /* if (readed == null) {
            throw new IllegalArgumentException("Product wasn't found");
        }*/if (readed.dtUpdate != dtUpdate) {
            throw OptimisticLockException(EDITED)
        }
        val productUpdate = storage.save(
            create()
                .setId(readed.id)
                .setDtCreate(readed.dtCreate)
                .setDtUpdate(Clock.systemUTC().millis())
                .setTitle(idto.title)
                .setWeight(idto.weight)
                .setCalories(idto.calories)
                .setProteins(idto.proteins)
                .setFats(idto.fats)
                .setCarbohydrates(idto.carbohydrates)
                .setCreatedByRole(login)
                .build()
        )
        auditService!!.create(user, EntityType.PRODUCT, UPDATED, productUpdate.id.toString())
        return productUpdate
    }

    override fun delete(product: Product) {}
    fun validate(idto: InputProductDTO?) {
        checkNotNull(idto) { "You didn't pass the product" }
        checkNotNull(idto.title) { "You didn't pass the Title" }
        check(!(idto.fats < 0 || idto.fats == 0.0)) { "You didn't pass the value of Fats" }
        check(!(idto.carbohydrates < 0 || idto.carbohydrates == 0.0)) { "You didn't pass the value of Carbohydrates" }
        check(!(idto.proteins < 0 || idto.proteins == 0.0)) { "You didn't pass the value of Proteins" }
        check(!(idto.calories < 0 || idto.calories == 0.0)) { "You didn't pass the value of Calories" }
        check(!(idto.weight < 0 || idto.weight == 0.0)) { "You didn't pass the value of Weight" }
    }
}